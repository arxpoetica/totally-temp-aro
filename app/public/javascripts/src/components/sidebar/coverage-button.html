<!-- Show the "Initialize" button only if we don't have a report -->
<button ng-if="!$ctrl.state.coverage.report"
        class="btn btn-block btn-primary"
        ng-disabled="$ctrl.isInitializingCoverage"
        ng-click="$ctrl.initializeCoverageReport()">
  <i class="fa fa-bolt"></i> Initialize coverage
</button>

<!-- Show the "Modify" button only if we have a report -->
<button ng-if="$ctrl.state.coverage.report"
        class="btn btn-block btn-primary modify-coverage-button"
        ng-disabled="$ctrl.isModifyingCoverage"
        ng-click="$ctrl.modifyCoverageReport()">
  Modify coverage
</button>

<!-- A spacer div -->
<div style="width: 100%; padding-bottom: 20px"></div>
