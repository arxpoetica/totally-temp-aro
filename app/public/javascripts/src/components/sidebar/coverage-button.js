import StateCoverage from '../../models/state-coverage'

class CoverageButtonController {
  constructor(state, $http, $timeout) {
    this.state = state
    this.$http = $http
    this.$timeout = $timeout
    this.isInitializingCoverage = false
    this.isModifyingCoverage = false
  }

  areInputsComplete() {

    if (this.state.coverage.initializationParams.coverageType === 'census_block') {
      return true // For census_blocks/Form477 inputs are complete even without selecting locationtypes
    }

    // Check if at least one location type is selected for optimization
    const isLocationInputValid = this.state.locationInputSelected('business') || this.state.locationInputSelected('household') || this.state.locationInputSelected('celltower')
    return isLocationInputValid
  }

  initializeCoverageReport() {
    // Format the coverage report that so it can be sent over to aro-service
    var requestBody = {
      coverageAnalysisRequest: angular.copy(this.state.coverage.initializationParams)
    }
    requestBody.coverageAnalysisRequest.planId = this.state.plan.getValue().id
    requestBody.coverageAnalysisRequest.projectTemplateId = this.state.loggedInUser.projectId
    requestBody.coverageAnalysisRequest.analysisSelectionMode = this.state.optimizationOptions.analysisSelectionMode
    requestBody.coverageAnalysisRequest.locationTypes = this.state.locationTypes.getValue()
                                                                                        .filter(item => item.checked)
                                                                                        .map(item => item.plannerKey)
    if (this.state.optimizationOptions.analysisSelectionMode === this.state.selectionModes.SELECTED_ANALYSIS_AREAS) {
      // If we have analysis areas selected, we can have exactly one analysis layer selected in the UI
      const visibleAnalysisLayers = this.state.boundaries.tileLayers.filter(item => item.visible && (item.type === 'analysis_layer'))
      if (visibleAnalysisLayers.length !== 1) {
        const errorMessage = 'You must have exactly one analysis layer selected to initialize the coverage report'
        swal({
          title: 'Analysis Layer error',
          text: errorMessage,
          type: 'error',
          closeOnConfirm: true
        })
        throw errorMessage
      }
      requestBody.coverageAnalysisRequest.analysisLayerId = visibleAnalysisLayers[0].analysisLayerId
    }
    this.isInitializingCoverage = true
    this.$http.post(`/service/coverage/report`, requestBody)
      .then((result) => {
        this.state.coverage.report = result.data
        return this.$http.post(`/service/coverage/report/${this.state.coverage.report.reportId}/init?user_id=${this.state.loggedInUser.id}`, {})
      })
      .then(() => this.$http.post(`/service/coverage/report/${this.state.coverage.report.reportId}/process?user_id=${this.state.loggedInUser.id}`, {}))
      .then(() => {
        this.isInitializingCoverage = false
        this.$timeout()
      })
      .catch(err => {
        console.error(err)
        this.isInitializingCoverage = false
        this.$timeout()
      })
  }

  modifyCoverageReport() {
    this.isModifyingCoverage = true
    this.$http.delete(`/service/coverage/report/${this.state.coverage.report.reportId}`)
      .then(result => {
        this.isModifyingCoverage = false
        StateCoverage.initializeCoverage(this.state, this.$http, this.$timeout)
        this.$timeout()
      })
      .catch(err => {
        this.isModifyingCoverage = false
        console.error(err)
        this.$timeout()
      })
  }
}

CoverageButtonController.$inject = ['state', '$http', '$timeout']

let coverageButton = {
  templateUrl: '/components/sidebar/coverage-button.html',
  bindings: {},
  controller: CoverageButtonController
}

export default coverageButton