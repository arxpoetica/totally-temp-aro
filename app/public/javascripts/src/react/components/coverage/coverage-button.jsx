import React from 'react'
import ImmutablePropTypes from 'react-immutable-proptypes'
import { PropTypes } from 'prop-types'
import reduxStore from '../../../redux-store'
import CoverageActions from '../coverage/coverage-actions'
import wrapComponentWithProvider from '../../common/provider-wrapped-component'
import CoverageStatusTypes from './constants'
import { SocketManager } from '../../../react/common/socket-manager'
import ProgressButton from '../common/progress-button.jsx'

export class CoverageButton extends ProgressButton {
  constructor (props) {
    super(props)

    // override 
    this.statusTypes = {
      UNINITIALIZED: CoverageStatusTypes.UNINITIALIZED,
      RUNNING: CoverageStatusTypes.RUNNING,
      FINISHED: CoverageStatusTypes.FINISHED
    }

    this.unsubscriber = SocketManager.subscribe('PROGRESS_MESSAGE_DATA', (progressData) => {
      if (progressData.data.processType === 'coverage') {
        this.props.setCoverageProgress(progressData.data)
      }
    })
  }

  renderFinishedButton () {
    return (
      <button className={'btn btn-block modify-coverage-button'} style={{ marginBottom: '10px' }}
        onClick={() => this.props.modifyCoverageReport(this.props.report.reportId)}>
        <i className='fa fa-edit' /> Modify
      </button>
    )
  }

  // override 
  onRun () {
    this.props.initializeCoverageReport(this.props.userId, this.props.planId, this.props.projectId, this.props.activeSelectionModeId,
      this.props.locationLayers.filter(item => item.checked).map(item => item.plannerKey),
      this.props.boundaryLayers, this.props.initializationParams)
  } 
  
  // override 
  onModify () {
    this.props.modifyCoverageReport(this.props.report.reportId)
  }

  componentWillUnmount () {
    this.unsubscriber()
  }
}

CoverageButton.propTypes = {
  status: PropTypes.string,
  progress: PropTypes.number,
  userId: PropTypes.number,
  planId: PropTypes.number,
  projectId: PropTypes.number,
  locationLayers: ImmutablePropTypes.list,
  boundaryLayers: ImmutablePropTypes.list,
  activeSelectionModeId: PropTypes.string,
  initializationParams: PropTypes.object,
  report: PropTypes.object
}

const mapStateToProps = (state) => {
  return {
    status: state.coverage.status,
    progress: state.coverage.progress,
    userId: state.user.loggedInUser.id,
    planId: state.plan.activePlan && state.plan.activePlan.id,
    projectId: state.user.loggedInUser.projectId,
    locationLayers: state.mapLayers.location,
    boundaryLayers: state.mapLayers.boundary,
    activeSelectionModeId: state.selection.activeSelectionMode.id,
    initializationParams: state.coverage.initializationParams,
    report: state.coverage.report
  }
}

const mapDispatchToProps = (dispatch, ownProps) => ({
  modifyCoverageReport: (reportId) => dispatch(CoverageActions.modifyCoverageReport(reportId)),
  initializeCoverageReport: (userId, planId, projectId, activeSelectionMode, locationTypes, boundaryLayers, initializationParams) => {
    dispatch(CoverageActions.initializeCoverageReport(userId, planId, projectId, activeSelectionMode, locationTypes,
      boundaryLayers, initializationParams))
  },
  setCoverageProgress: (progress) => dispatch(CoverageActions.setCoverageProgress(progress))
})

const CoverageButtonComponent = wrapComponentWithProvider(reduxStore, CoverageButton, mapStateToProps, mapDispatchToProps)
export default CoverageButtonComponent
