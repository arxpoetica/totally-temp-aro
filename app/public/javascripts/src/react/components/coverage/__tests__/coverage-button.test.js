import React from 'react'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import { CoverageButton } from '../coverage-button'
import CoverageStatusTypes from '../constants'
import { List } from 'immutable';

// -----------------------------------------------------------------------------
test('When the user does not have a coverage report', () => {
  const component = renderer.create(
    <CoverageButton status={CoverageStatusTypes.UNINITIALIZED}></CoverageButton>
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

// -----------------------------------------------------------------------------
test('When the user has a coverage report', () => {
  const component = renderer.create(
    <CoverageButton status={CoverageStatusTypes.FINISHED}></CoverageButton>
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

// -----------------------------------------------------------------------------
test('When a coverage report is running at 10% progress', () => {
  const component = renderer.create(
    <CoverageButton status={CoverageStatusTypes.RUNNING} progress={0.1}></CoverageButton>
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

// -----------------------------------------------------------------------------
test('When button is clicked to initialize report', () => {
  const locationLayers = new List([
    { plannerKey: 'small', checked: true },
    { plannerKey: 'medium', checked: false }
  ])
  const boundaryLayers = new List([
    { key: 'analysis_area', name: 'Analysis Area', checked: true },
    { key: 'analysis_area', name: 'Analysis Area 2', checked: false }
  ])
  const mockInitializeCoverageReport = jest.fn()
  const component = shallow(
    <CoverageButton status={CoverageStatusTypes.UNINITIALIZED}
                    userId={5}
                    planId={22}
                    projectId={7}
                    locationLayers={locationLayers}
                    boundaryLayers={boundaryLayers}
                    activeSelectionModeId={'SELECTED_AREAS'}
                    initializationParams={{ generic: 'Init Params'}}
                    initializeCoverageReport={mockInitializeCoverageReport}></CoverageButton>
  )

  component.find('button').simulate('click')
  // We cannot use expect(mockInitializeCoverageReport).toHaveBeenCalledWith() because of the immutable list
  expect(mockInitializeCoverageReport.mock.calls[0][0]).toEqual(5)
  expect(mockInitializeCoverageReport.mock.calls[0][1]).toEqual(22)
  expect(mockInitializeCoverageReport.mock.calls[0][2]).toEqual(7)
  expect(mockInitializeCoverageReport.mock.calls[0][3]).toEqual('SELECTED_AREAS')

  const expectedLocationLayers = locationLayers.filter(item => item.checked).map(item => item.plannerKey)
  expect(mockInitializeCoverageReport.mock.calls[0][4].equals(expectedLocationLayers)).toBeTruthy()

  expect(mockInitializeCoverageReport.mock.calls[0][5].equals(boundaryLayers)).toBeTruthy()
  expect(mockInitializeCoverageReport.mock.calls[0][6]).toEqual({ generic: 'Init Params'})
})

// -----------------------------------------------------------------------------
test('When button is clicked to modify/delete report', () => {
  const mockModifyCoverageReport = jest.fn()
  const component = shallow(
    <CoverageButton status={CoverageStatusTypes.FINISHED}
                    report={{ reportId: 100 }}
                    modifyCoverageReport={mockModifyCoverageReport}>
    </CoverageButton>
  )

  component.find('button').simulate('click')
  expect(mockModifyCoverageReport).toHaveBeenCalledWith(100)
})
