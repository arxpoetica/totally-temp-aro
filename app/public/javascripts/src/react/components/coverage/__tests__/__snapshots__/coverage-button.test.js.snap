// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When a coverage report is running at 10% progress 1`] = `
<div
  className="progress"
  style={
    Object {
      "height": "34px",
      "marginBottom": "10px",
      "position": "relative",
    }
  }
>
  <div
    aria-valuemax="1"
    aria-valuemin="0"
    aria-valuenow={0.1}
    className="progress-bar progress-bar-optimization"
    role="progressbar"
    style={
      Object {
        "lineHeight": "34px",
        "width": "10%",
      }
    }
  />
  <div
    style={
      Object {
        "backgroundColor": "rgba(0, 0, 0, 0.4)",
        "borderRadius": "3px",
        "color": "white",
        "fontWeight": "bold",
        "left": "50%",
        "position": "absolute",
        "textAlign": "center",
        "top": "50%",
        "transform": "translateX(-50%) translateY(-50%)",
        "width": "80px",
      }
    }
  >
    10%
  </div>
</div>
`;

exports[`When the user does not have a coverage report 1`] = `
<button
  className="btn btn-block btn-primary"
  disabled={false}
  onClick={[Function]}
  style={
    Object {
      "marginBottom": "10px",
    }
  }
>
  <i
    className="fa fa-bolt"
  />
   Run
</button>
`;

exports[`When the user has a coverage report 1`] = `
<button
  className="btn btn-block modify-coverage-button"
  disabled={false}
  onClick={[Function]}
  style={
    Object {
      "marginBottom": "10px",
    }
  }
>
  <i
    className="fa fa-edit"
  />
   Modify
</button>
`;
